pyannote/database/loader.py

```diff
@@ -187,6 +187,47 @@ def __call__(self, file: ProtocolFile) -> Timeline:
         return self.loaded_[uri]
 
 
+class LabLoader:
+    """LAB loader
+
+    Parameters
+    ----------
+    path : str
+        Path to LAB file with ProtocolFile
+        (e.g. "/path/to/{uri}.lab")
+
+        each .lab file contains the segments for a single audio file, in the following format:
+        start end label
+
+        ex.
+        0.0 12.3456 sing
+        12.3456 15.0 nosing
+        ...
+    """
+
+    def __init__(self, path: Text = None):
+        super().__init__()
+
+        self.path = str(path)
+        self.data_ = load_lab(self.path)
```

Comment:
Since `path` contains a placeholder, it cannot be loaded at `__init__` time -- loading must be done in `__call__`.

You can take inspiration from RTTMLoader placeholder handling to do the same in LABLoader.

Reply:
I rewrote all the function looking at the RTTMLoader:
`
class LabLoader:
    """LAB loader

    Parameters
    ----------
    path : str
        Path to LAB file with ProtocolFile
        (e.g. "/path/to/{uri}.lab")
    """

    def __init__(self, path: Text = None):
        super().__init__()

        self.path = str(path)
        
        _, placeholders, _, _ = zip(*string.Formatter().parse(self.path))
        self.placeholders_ = set(placeholders) - set([None])
        self.loaded_ = dict() if self.placeholders_ else load_lab(self.path)
        
    def __call__(self, file: ProtocolFile) -> Annotation:

        uri = file["uri"]

        if uri in self.loaded_:
            return self.loaded_[uri]

        sub_file = {key: file[key] for key in self.placeholders_}
        loaded = load_lab(self.path.format(**sub_file), uri)
        if uri not in loaded:
            loaded[uri] = Timeline(uri=uri)
        
        # do not cache timelines when there is one LAB file per "uri"
        # since loading it should be quite fast
        if "uri" in self.placeholders_:
            return loaded[uri]

        # when there is more than one file in loaded LAB, cache them all
        # so that loading future "uri" will be instantaneous
        self.loaded_.update(loaded)

        return self.loaded_[uri]
`

but I didn't understood how to use the placeholder processing. What should I write in the database.yml file? Basically it runs the load_lab in the init, because the placeholder var is empty.

Reply:
The `database.yml` should actually contain the `"{uri}"` string. 
It is replaced automatically in the `__call__` method by the content of `file["uri"]`.
See https://github.com/pyannote/pyannote-database/#data-loaders